### PR TITLE
Feature app

### PR DESCRIPTION
**Pull request by sccolbert**
Aside from updating tests, and a once over of the docs, this should be g2g and is what we need to move JLab over to the monorepo.

### CODE REVIEW

**Comment by sccolbert** [comment]
Major enhancements:
- Application object which supports plugins
- Virtual DOM
- TabBar updates to support vertical tabs

Differences from Phosphide:
- No more pre-canned application shell widget. JLab will have to make it's own shell (just take the one from Phosphide and tweak as needed).

File: src/ui/tabbar.ts [hunk 1]
```diff
@@ -1125,102 +1338,150 @@ namespace TabBar {
...
-      let tabInfix = title.className ? ` ${title.className}` : '';
-      let tabSuffix = title.closable ? ` ${CLOSABLE_CLASS}` : '';
-      let iconSuffix = title.icon ? ` ${title.icon}` : '';
-      let icon = node.firstChild as HTMLElement;
-      let label = icon.nextSibling as HTMLElement;
-      node.className = `${TAB_CLASS} ${tabInfix} ${tabSuffix}`;
-      icon.className = `${ICON_CLASS} ${iconSuffix}`;
-      label.textContent = title.label;
-      label.title = title.caption;
+    createTabclass(data: IRenderData): string {
```
Comment (blink1073) [inline src/ui/tabbar.ts]: `createTabClass` (camelCase)?

File: src/ui/vdom.ts [hunk 2]
```diff
@@ -0,0 +1,1481 @@
...
+ * @returns A new virtual DOM node for the given parameters.
+ *
+ * #### Notes
+ * The children may be string literals, other virtual DOM nodes, or an
+ * array of either of those things. String literals are converted into
+ * text nodes, and arrays are inlined as if their contents were given
+ * as positional arguments. This makes it easy to build up an array of
+ * children by any desired means.
+ *
+ * A strongly typed bound function for each tag name is aslo available
```
Comment (blink1073) [inline src/ui/vdom.ts]: *also

**Comment by sccolbert** [comment]
The tests pass locally. Travis is timing out though (typical).

File: src/ui/vdom.ts [hunk 3]
```diff
@@ -0,0 +1,1481 @@
...
+ */
+export
+interface IVNodeAttrs {
+  /**
+   * The key id for the node.
+   *
+   * If a node is given a key id, the generated DOM node will not be
+   * recreated during a rendering update if it only moves among its
+   * siblings in the render tree, provided the type of the node does
+   * not change.
```
Comment (afshin) [inline src/ui/vdom.ts]: Perhaps explain why somebody might want this:

> This behavior is an optimization for nodes whose children are known to be static. One example of this usage is the nodes generated for tabs in a tab bar: while their relative order amongst their siblings may change, their content (child nodes) will not change, so they can be implemented as keyed nodes.

Or something of that nature.

**Comment by blink1073** [comment]
👍

**Comment by sccolbert** [comment]
TIL github won't let you merge failing pull requests from a phone. This is only failing cause Travis is dumb. I'll merge when I'm back on my laptop.

**Comment by blink1073** [comment]
Ha, and I don't have permissions either.

**Comment by jasongrout** [comment]
Scroll down to the bottom and choose desktop version at the bottom, if you're looking at the GitHub website.

**Comment by sccolbert** [comment]
Werd

**Comment by blink1073** [comment]
bigyes

**Pull request merged**